JCLOUDS-1386  1&1 Baremetal servers

alibazlamit opened the pull request:
This PR holds the latest updates in the 1&1 API

alibazlamit commented:
Compute test results :
https://gist.github.com/alibazlamit/fab05a5e3235be8202f34c627d8bdc65

andreaturli commented:
thanks @alibazlamit could you please open a jira issue that describes the rationale for this PR?
Thanks

alibazlamit commented:
@andreaturli one main feature that this PR is introducing is the Bare-metal Servers, now i have implemented that as a feature but it is not integrated with the compute API. Basically there is a new property introduced `server_type` which defaults to `cloud` or could be `baremetal`, do you have any idea/examples where you can choose the server type when building a template, i know that with templates you can choose which images to use and what hardware goes, but how can we introduce choosing a `baremetal` or `cloud` server type?

Thanks

andreaturli commented:
Thanks @alibazlamit for the summary.

Would you mind to copy that into a jira issue as described at point 2 at https://cwiki.apache.org/confluence/display/JCLOUDS/How+to+Contribute
Ideally you can rename the PR using the jira issue I'd.

That would hugely help us to manage the project.Thanks

alibazlamit commented:
I am sorry for the confusion, but there was a question in my previous comment, something that i would like to add to the PR before we merge it. Could you please take a look at it and let me know if you can maybe point me towards a solution.

Thanks

andreaturli commented:
Not familiar with the oneandone API but looks like `baremetal` is another servertype for server creation.

If this is correct, then we could either introduce a OneAndOneTemplateOption for serverType or try to infer it from hardwareProfiles if this additional flag is available. Makes sense?

alibazlamit commented:
I have implemented the baremetal creation in the compute service adapter, based on the hardware flavor chosen we can determine the server type.

Here is the [JIRA ](https://issues.apache.org/jira/browse/JCLOUDS-1386)ticket.

alibazlamit commented:
@andreaturli a friendly reminder, this is a review ready PR

andreaturli commented:
Thanks @alibazlamit, we'll have asap. Thanks

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -65,6 +66,7 @@
    private final OneAndOneApi api;
    private final Predicate<Server> waitServerUntilAvailable;
    private final PasswordGenerator.Config passwordGenerator;
```
it seems not used

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -83,56 +85,112 @@
-
+      String imageId = image.getId();
+      HardwareFlavour hardwareModel = isFlavor(hardware.getId());
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
```
I think worth introducing a `OneandoneTemplateOptions extends TemplateOptions` to deal with specific requirements a user may specify like bare metal or even `hardwareModel` above. wdyt?

alibazlamit replied:
My intention was to allow the users to just pass in the flavor name i.e. `BMC_XL` or `XL` as a `hardwareId` to choose between baremetal and cloud servers.

Maybe the `OneandoneTemplateOptions` would be an overkill since the only thing users will need to provide is the different hardware flavor name which can be achieved using the `public TemplateBuilder hardwareId(String string);` from the `TeamplateBuilder` class

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -83,56 +85,112 @@
       try {
-         org.apache.jclouds.oneandone.rest.domain.Hardware.CreateHardware hardwareRequest
-                 = org.apache.jclouds.oneandone.rest.domain.Hardware.CreateHardware.create(cores, 1, ram, hdds);
+         if (hardwareModel != null) {
```
this logic to build the hardwareRequest is quite complex. would it make sense to extract it in its own method, and add some unit tests, ideally

alibazlamit replied:
done [here](https://github.com/StackPointCloud/jclouds-labs/blob/oneandone-baremetal-update/oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/GenerateHardwareRequest.java)

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -179,9 +220,54 @@
+         result.add(flavor);
+      }
+      List<BareMetalModel> baremetalModels = api.serverApi().listBaremetalModels();
+      for (BareMetalModel model : baremetalModels) {
```
maybe you can consider a `Function<BareMetalModel, HardwareFlavour>` for better readability and testability?

alibazlamit replied:
I need to combine `BareMetalModel` with `HardwareFlavour` as one list, to mix the baremetal flavors with the cloud flavors, do you think that could be achieved using the function you suggested?

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/config/OneAndOneComputeServiceContextModule.java:
```diff
@@ -125,4 +126,28 @@ public boolean apply(Server server) {
+
+   @Provides
+   @Singleton
+   @Named(POLL_PREDICATE_SERVER_SUSPENDED)
```
why a not using `ComputeServiceProperties.TIMEOUT_NODE_SUSPENDED` so that you can use a quite adopted pattern like in openstack,digitalocean,packet
```
   @Provides
   @com.google.inject.name.Named(TIMEOUT_NODE_TERMINATED)
   protected Predicate<RegionAndId> provideServerTerminatedPredicate(final NovaApi api, ComputeServiceConstants.Timeouts timeouts,
                                                                ComputeServiceConstants.PollPeriod pollPeriod) {
      return retry(new ServerTerminatedPredicate(api), timeouts.nodeTerminated, pollPeriod.pollInitialPeriod,
              pollPeriod.pollMaxPeriod);
   }
```
?

andreaturli replied:
@alibazlamit why do we need this `POLL_PREDICATE_SERVER_SUSPENDED` when jclouds has `TIMEOUT_NODE_SUSPENDED` ?

alibazlamit replied:
I have changed it to `TIMEOUT_NODE_SUSPENDED`

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/HardwareFlavourToHardware.java:
```diff
@@ -53,7 +53,7 @@ public Hardware apply(HardwareFlavour from) {
       final HardwareBuilder builder;
       builder = new HardwareBuilder()
-              .ids(from.id())
+              .ids(from.name())
```
maybe better to use 
```
.id(from.id())
.name(from.name())
```
if `.ids()` is not what you want here

alibazlamit replied:
My first approach was what you have suggested, for some reason the API has similiar Id's between the baremetal flavors and the cloud flavors, the only thing to make them unique is the name so far, this might be an issue on the API side that might get changed in the future, if that ever happens we can go with what you suggested

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/config/OneAndOneProperties.java:
```diff
@@ -19,6 +19,7 @@
 public final class OneAndOneProperties {
 
    public static final String POLL_PREDICATE_SERVER = "jclouds.oneandone.rest.predicate.server";
+   public static final String POLL_PREDICATE_SERVER_SUSPENDED = "jclouds.oneandone.rest.predicate.serverdelete";
```
oh now I see, it's a predicate to wait for server deletion not really suspension. ok maybe my comment above is still valid for other predicates though

alibazlamit replied:
I don't quite get what is required here could you please explain more?

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/BareMetalModel.java:
```diff
@@ -0,0 +1,70 @@
+
+      public abstract String unit();
+
+      public abstract Hdd hdds();
```
out of curiosity, why it's `hdds` if it is not a collection?

alibazlamit commented:
@andreaturli Thanks for the review, i have applied most of the changes you requested.
I wanted to run the live tests, but four of them fail on when trying to ssh into the server, for example `testCreateAndRunAService` fails with this error 
```
(web:rsa[fingerprint(37:18:5e:f4:ce:d4:95:c4:02:10:8c:01:b9:b9:d3:4b),sha1(c0:81:47:7b:a1:88:47:2d:e0:1c:e4:6a:28:d6:fc:ab:c3:6b:44:16)]@50.21.186.61:22) (web:rsa[fingerprint(37:18:5e:f4:ce:d4:95:c4:02:10:8c:01:b9:b9:d3:4b),sha1(c0:81:47:7b:a1:88:47:2d:e0:1c:e4:6a:28:d6:fc:ab:c3:6b:44:16)]@50.21.186.61:22) error acquiring {hostAndPort=50.21.186.61:22, loginUser=web, ssh=null, connectTimeout=300000, sessionTimeout=300000} (out of retries - max 7): Exhausted available authentication methods
org.jclouds.rest.AuthorizationException
	at org.jclouds.sshj.SshjSshClient.propagate(SshjSshClient.java:394)
```

I am not sure why is it trying to ssh into the server with loginUser=web, there is not such user at the created server and it is not going to be able to ssh, any ideas?

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
    public NodeAndInitialCredentials<Server> createNodeWithGroupEncodedIntoName(String group, String name, Template template) {
 
-      final String dataCenterId = template.getLocation().getId();
+      final String dataCenterId ="908DC2072407C94C8054610AD5A53B8C";// template.getLocation().getId();
```
why this hardcoded value?

alibazlamit replied:
Mistake, fixed.

alibazlamit replied:
Mistake, fixed.

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
    OneandoneComputeServiceAdapter(OneAndOneApi api, CleanupResources cleanupResources,
+           GenerateHardwareRequest generateHardwareRequest,
            @Named(POLL_PREDICATE_SERVER) Predicate<Server> waitServerUntilAvailable,
            PasswordGenerator.Config passwordGenerator) {
```
as you have removed it from the constructor, why do you still need to inject it?

alibazlamit replied:
It is still in the constructor as i need to use it.

andreaturli replied:
`PasswordGenerator.Config passwordGenerator` is not used, remove it

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
+      final String dataCenterId ="908DC2072407C94C8054610AD5A53B8C";// template.getLocation().getId();
       Hardware hardware = template.getHardware();
       TemplateOptions options = template.getOptions();
       Server updateServer = null;
```
null not needed, maybe remove it?

alibazlamit replied:
removed

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
+      String imageId = image.getId();
+      Hardware hardwareModel = generateHardwareRequest.isFlavor(hardware.getId());
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
+      ServerAppliance workingImage = null;
```
null not needed, maybe remove it?

alibazlamit replied:
removed

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
+      }
+      String firewallPolicyId = "";
+      if (inboundPorts.length > 0) {
+         FirewallPolicy firewallPolicy = api.firewallPolicyApi().create(FirewallPolicy.CreateFirewallPolicy.create(name + " firewall policy", "desc", rules));
```
what happens when you ask to create multiple nodes? does `firewallPolicyApi` handles the concurrency correctly?

jclouds usually takes care of fw rules for a group of node extending `CreateNodesWithGroupEncodedIntoNameThenAddToSet` -- see `ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSet` in openstack-nova api for an example

alibazlamit replied:
It does handle the creation correctly as all the live tests passes, if you think there is a necessary change here, we can do that in another PR that addresses improvements.

This bulk of code was moved up just to organize the process more. it has been merged before since the first PR

andreaturli replied:
I've now realised I already commented this section: I'm afraid you need to change the code now otherwise we can't merge this PR. The reason for that is to have homogenous code, easier to maintain and debug.
if each contributions start deviating from jclouds's best practices it would become impossible for maintainers to help out. I'm sure you understand

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
+      ServerAppliance workingImage = null;
+
+      //choose the correct image based on the server type baremetal or cloud
```
sorry can't understand this.

Can you elaborate a bit more the rationale of this `findWorkingImage` method?

alibazlamit replied:
Some of the images are of Type ISO and type Image, the ones that are of type ISO are basically an installation of the Linux system and not needed here, this function makes sure to choose the correct boot-able image based on the image from the template at hand

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -62,77 +61,95 @@
-      } else {
-         ram = ram / 1024;
+      //configuring Firewall rules
+      Map<Integer, Integer> portsRange = getPortRangesFromList(inboundPorts);
```
Is this change strictly related to bare metal support? seems to me that this PR is trying to change too many things, would it be possible for you to split up into multiple smaller PR so to have a better/easier review?

alibazlamit replied:
This has just been moved to the GenerateHardwareRequest.java to organize code just as you request to separate the creation of the hardware request.

nacx commented:
>I am not sure why is it trying to ssh into the server with loginUser=web, there is not such user at the created server and it is not going to be able to ssh, any ideas?

There is a test that installs a Jetty Server and creates an admin ("web") user for that service to test end-to-end a typical use case. It is using that "web" user to try access your node. If the login fails, it may be due to the previous script failing or because the user could not be properly configured in the OS. You could check that.
Anyway, let's first address Andrea's comments and once everything is clean we can investigate this failure.

alibazlamit commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -143,23 +160,6 @@
 
          updateServer = api.serverApi().get(server.id());
 
-         Map<Integer, Integer> portsRange = getPortRangesFromList(inboundPorts);
```
This bulk of code was moved up

alibazlamit commented:
> andreaturli commented on Feb 14  
Not familiar with the oneandone API but looks like baremetal is another servertype for server creation.
> 
> If this is correct, then we could either introduce a OneAndOneTemplateOption for serverType or try to infer it from hardwareProfiles if this additional flag is available. Makes sense?

@andreaturli I am looking into that option just in case, could you please point me to an example where one of the current providers actually use and TemplateOption, i would like to investigate that and see if it would fit the case here.

alibazlamit commented:
@andreaturli @nacx just a friendly reminder, i have addresses all of the comments either by applying the appropriate changes or be replying to them here, could we please move on with this PR?

alibazlamit commented:
@andreaturli Can you please take a look at this PR and let me know when can we move on with it?

alibazlamit commented:
@andreaturli Could you please let me know which issues are not addressed, i replied to every you comment you had, some changes are applied the others are explained as you can see.

andreaturli commented:
@alibazlamit I've added a question to help the discussion. Can we start from there?

alibazlamit commented:
@andreaturli  I have addressed the issues you mentioned, could you now please take a look?

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Types.java:
```diff
@@ -253,7 +253,7 @@ public static ApplianceType fromValue(String v) {
 
    public enum OSImageType {
-      Standard, Minimal, Personal, ISO_OS, ISO_TOOL, NULL, UNRECOGNIZED;
+     STANDARD, MINIMAL, Personal, ISO_OS, ISO_TOOL, NULL, UNRECOGNIZED;
```
please make it all capital

alibazlamit replied:
done

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Types.java:
```diff
@@ -283,4 +283,22 @@ public String toString() {
+   }
+
+   public enum ServerTypeCompatibility {
+      vps, cloud, baremetal, UNRECOGNIZED;
```
upper case?

alibazlamit replied:
done

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Types.java:
```diff
@@ -283,4 +283,22 @@ public String toString() {
    }
+
+   public enum ServerType {
+      cloud, baremetal, UNRECOGNIZED;
```
why not upper case?

alibazlamit replied:
done

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -179,9 +179,46 @@
       return new NodeAndInitialCredentials<Server>(updateServer, updateServer.id(), serverCredentials);
    }
 
+   private ServerAppliance findWorkingImage(String imageId, Types.ServerTypeCompatibility serverType) {
```
why this name? it seems that it tries to find a `ServerAppliance`, doesn't it?

alibazlamit replied:
changed

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -82,57 +83,73 @@
-
+      String imageId = image.getId();
+      Hardware hardwareModel = generateHardwareRequest.isFlavor(hardware.getId());
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
```
wouldn't be better to introduce a custom `TemplateOptions` where you can pass in `bareMetal` boolean parameter? see https://github.com/jclouds/jclouds/blob/master/providers/digitalocean2/src/main/java/org/jclouds/digitalocean2/compute/options/DigitalOcean2TemplateOptions.java#L38 for an example

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -82,57 +83,73 @@
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
+      ServerAppliance workingImage;
+
+      //choose the correct image based on the server type baremetal or cloud
```
move this business logic into its own method

alibazlamit replied:
no need as the PR will get even more complicated to review

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -82,57 +83,73 @@
-      } else {
-         ram = ram / 1024;
+      //configuring Firewall rules
+      Map<Integer, Integer> portsRange = getPortRangesFromList(inboundPorts);
```
all the reasonings for the firewalls / inbound ports shouldn't probably happen in the adapter: firewalls are created once and attached to all the nodes of a group. See `CreateResourcesThenCreateNodes` as a reference at https://github.com/jclouds/jclouds/blob/master/providers/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/strategy/CreateResourcesThenCreateNodes.java#L138

alibazlamit replied:
I suggest we open a separate PR addressing any backwards concern you might have

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -259,19 +289,25 @@ public void destroyNode(String id) {
    @Override
    public void resumeNode(String id) {
-      api.serverApi().updateStatus(id, Server.UpdateStatus.create(Types.ServerAction.POWER_ON, Types.ServerActionMethod.HARDWARE));
+      api.serverApi().updateStatus(id, Server.UpdateStatus.create(Types.ServerAction.POWER_ON, Types.ServerActionMethod.HARDWARE, false, null));
```
what happens if the node is already started?

alibazlamit replied:
nothing happens

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -259,19 +289,25 @@ public void destroyNode(String id) {
    public void suspendNode(String id) {
       waitServerUntilAvailable.apply(getNode(id));
-      api.serverApi().updateStatus(id, Server.UpdateStatus.create(Types.ServerAction.POWER_OFF, Types.ServerActionMethod.HARDWARE));
+      api.serverApi().updateStatus(id, Server.UpdateStatus.create(Types.ServerAction.POWER_OFF, Types.ServerActionMethod.HARDWARE, false, null));
```
what if the node is already powered off?

alibazlamit replied:
it will remain powered off

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -78,26 +81,43 @@ public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
+      String hardwareId = null;
 
       DataCenter dataCenter = api.dataCenterApi().get(server.datacenter().id());
       Location location = find(locations.get(), idEquals(dataCenter.id()));
```
you already have the `datacenter.id` you don't need to call the api to get the datacenter object

alibazlamit replied:
changed

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -78,26 +81,43 @@ public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
       //check if the server was built on a hardware flavour(Fixed instance)
-      if (server.hardware().fixedInstanceSizeId() != null && !"0".equals(server.hardware().fixedInstanceSizeId())) {
-         hardware = hardwareFlavors.get().get(server.hardware().fixedInstanceSizeId());
+      if (server.serverType() == Types.ServerType.baremetal) {
```
reuse `HardwareFlavourToHardware` to get the same information, but without calling the `api` again

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -78,26 +81,43 @@ public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
-         size = getHddSize(hdds);
+         size = (float) getHddSize(hdds);
+         if (size == 0) {
+            Server refreshedServer = api.serverApi().get(server.id());
```
you should avoid the extra `api` calls inside `ServerToNodeMetadata` - please remove `api` injection and refactor this class to not use it

andreaturli commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/GenerateHardwareRequest.java:
```diff
@@ -0,0 +1,103 @@
+import org.jclouds.compute.util.ComputeServiceUtils;
+
+@Singleton
+public class GenerateHardwareRequest {
```
can you add some javadoc for this class? I don't understand its usage

alibazlamit commented:
@andreaturli a few of your comments do make sense about this PR, i will address those issues so we can move on.

I don't see why i would separate Recovery Images it is a very small feature and i am not sure that will really affect reviewing the PR. In my opinion i think that most of the things that you are asking me to change will make reviewing this PR difficult, like changing firewall policies and moving few lines of code into a separate method, changes like that made this PR look messy and increase the diff.

I agree with you about being committed to jClouds best practices, but from my experience this is not the right PR to go back and review things that were already accepted by jClouds.

I have addressed all the small changes that make sense, i can post the live test results yet again.I hope you agree with me about closing this PR before its gets even worst
